### ISSUE TITLE
Why do we need FindEIGEN3.cmake ?

### ISSUE THREAD

**Issue by joxoby** [ignitionrobotics/ign-cmake#119]
I'm having a hard time understanding why do we need `FindEIGEN3.cmake`

Can't we rely on the standard CMake lookup process and just do `find_package(Eigen3 CONFIG)` ?

`Eigen3` comes with an `Eigen3Config.cmake` file.

By the way, I'm asking this question because I'm trying to add `ignition-math` as a Conan package: https://github.com/conan-io/conan-center-index/pull/3215

**Comment by mxgrey** [comment]
If I recall correctly, it was only (relatively) recently that Eigen3 started providing a cmake config-file. The official debian package for Eigen3 on one of the earlier platforms that we needed to support did not provide the config-file.

I don't remember which platform was missing the config-file, but if we no longer support it, then we can consider getting rid of the `FindEIGEN3.cmake` find-module.

**Comment by traversaro** [comment]
> I don't remember which platform was missing the config-file, but if we no longer support that platform, then we can consider getting rid of the FindEIGEN3.cmake find-module.

I guess the latest Ubuntu LTS without an Eigen3Config.cmake was Ubuntu 14.04 Trusty, see https://github.com/ros/cmake_modules/issues/25 . However, if we drop the `FindEIGEN3.cmake` , where will be the pkg-config-related function `ign_pkg_config_entry` called?

**Comment by joxoby** [comment]
@traversaro why do we need any `pkg-config` functionality?

**Comment by traversaro** [comment]
All the ignition libraries generate `.pc` files, and so they need to know the `pkg-config` info of their dependencies, so they can generate correctly the `Requires:` line of the generated `.pc` . I do not need this functionality, but this is what is implemented now.

**Comment by joxoby** [comment]
I see. I wonder who is using the `pkg-config` feature or why do we support it.

**Comment by mxgrey** [comment]
Great point about pkg-config. The `ign_find_package(~)` command [has some arguments to provide it manually](https://github.com/ignitionrobotics/ign-cmake/blob/a4c8568a4ef91b0abc6c64a50185721f2fd1a33d/cmake/IgnUtils.cmake#L104-L134), but it really is simpler to keep it all in one place.

Is there any problem with the `FindEIGEN.cmake` file that would motivate us to remove it? If it's causing collisions with other find-modules, then we could consider renaming it to something like `FindIgnEigen.cmake`.

**Comment by mxgrey** [comment]
> I see. I wonder who is using the pkg-config feature or why do we support it.

We definitely have users who prefer pkg-config over cmake because pkg-config is considered more portable between different build systems, and not all of our users like to use cmake.

**Comment by joxoby** [comment]
> Is there any problem with the FindEIGEN.cmake file that would motivate us to remove it? If it's causing collisions with other find-modules, then we could consider renaming it to something like FindIgnEigen.cmake.

See: https://github.com/conan-io/conan-center-index/pull/3268#issuecomment-716830374

**Comment by mxgrey** [comment]
I don't know what that reviewer's concern is. Packaging and distributing cmake find-modules for non-cmake dependencies is considered good practice.

The only issue I can think of is if our find-modules conflict with another package's find-modules and someone wants to use both our package and the other package as dependencies for the same project, and the find-modules exhibit critically different behaviors.

**Comment by joxoby** [comment]
I think that the question here is "Why are these `FindXXX.cmake` files part of `ignition-cmake` and not part of `ignition-<lib>`?"

The reviewer's concern is that when packaging `ignition-<lib>`, `FindXXX.cmake` can fail, in which case one would have to patch and repackage `ignition-cmake`.

`ignition-cmake` seems perfectly suitable to store macros that are used across all the different libraries. However, I'm not convinced that it should contain `.cmake` scripts to find libraries that are specific to each ignition project.

**Comment by mxgrey** [comment]
> The reviewer's concern is that when packaging ignition-<lib>, FindXXX.cmake can fail

This is what I'm lost on. What does it mean for `FindXXX.cmake` to fail? Does it mean calling `find_package(XXX)` from cmake fails to find package `XXX` or is there some other meaning of failure that I'm not thinking of?

If that's what's meant by fail, then that's perfectly normal. find-modules are designed to fail gracefully when the desired package is not available. Cmake itself is distributed with lots of find-modules for packages that users might not actually have, like boost.

I wonder if there may be some confusion here between the nature of a find-module and a config-file. Config-files are meant exclusively for packages that are definitely installed, but that's not what we're distributing here. We're distributing find-modules which are redistributable cmake modules whose job is to sniff around for the presence of a package.

I can agree that there's a fair debate to be had about whether the "best" place for these modules would be `ignition-cmake` or if it would be the respective ignition libraries that depend on them. The decision to put them all in `ignition-cmake` was based on primarily these factors:

1. It's easier to maintain all the heavy lifting related to cmake in one repo. The downstream packages don't need to do anything special related to cmake except declare their dependencies and name their libraries.
2. If multiple downstream packages share a dependency, we only need to have one copy of the find-module for it, here in this repo.
3. We only need to worry about putting one path into `CMAKE_MODULE_PATH`. If every downstream package hosted its own find-modules, that variable could get bloated quickly. This isn't a terrible problem, but I felt it would be cleaner this way.

The only downside I can think of for putting the find-modules in `ignition-cmake` is that we need to bump the minor version of this package any time a downstream package takes on a new dependency. This happens rarely enough that it's never been a bother for us.

**Comment by mxgrey** [comment]
tl;dr: find-modules are just another type of cmake module, just like the cmake modules that provide functions or macros. They don't have any packaging dependencies or requirements besides cmake.

**Comment by scpeters** [comment]
this conversation seems to have reached a stopping point. I'll close for now; please reopen if there are still concerns
